### PR TITLE
Use a pull model to withdraw deposit ETH balances

### LINKED ISSUE #551: Secure payments from bad recipient behavior
There are several cases where a third party is the recipient of an ETH payment from tBTC:
- Liquidation sends the appropriate amount of seized bonds to the liquidation auction winner, and potentially returns the leftover bond to the liquidation initiator, and possibly to the signers.
- Fraud liquidation during redemption immediately sends the redeemer the seized bonds from the signing group.
- Signing group fraud during funding sends the funder the seized bonds from the signing group.
- Signing group setup failure refunds the funder for the deposit opening fee and potentially returns leftover bond to the signers.
- Successful redemption returns leftover bond to the signers.

In all these cases, it's important that a misconfigured or malicious contract on the receiving side cannot prevent the system from completing the action it is executing.

There are two approaches here: additional bookkeeping and allowing relevant parties to *withdraw* their due amounts, or securing the outbound transfers so that they cannot break the remainder of the transaction.

### PR DESCRIPTION
**Pull request by NicholasDotSol**
Push models introduce security and stability concerns, a pull model allows for better security, as well as more flexibility and freedom to third-party contract fallbacks given `.call.value()`'s gas-forwarding.

- Accumulated Eth balance can only be withdrawn after the contract reaches an end-state.

Closes #551.

### CODE REVIEW

File: solidity/contracts/deposit/DepositFunding.sol [hunk 1]
```diff
@@ -169,7 +162,11 @@ library DepositFunding {
         bool _isFraud = _d.submitSignatureFraud(_v, _r, _s, _signedDigest, _preimage);
         require(_isFraud, "Signature is not fraudulent");
         _d.logFraudDuringSetup();
-        distributeSignerBondsToFunder(_d);
+
+        // allow deposit owner to withdraw sized bonds after contract termination;
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/DepositFunding.sol:166]: Seized? Guessing this is meant to be capitalized + end with a period as well <_<

File: solidity/contracts/deposit/DepositLiquidation.sol [hunk 2]
```diff
@@ -195,16 +195,15 @@ library DepositLiquidation {
         if (initiator == address(0)){
             initiator = address(0xdead);
         }
-        if (contractEthBalance > 1) {
+        if (contractEthBalance > valueToDistribute) {
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/DepositLiquidation.sol:198]: + 1?

File: solidity/contracts/deposit/DepositUtils.sol [hunk 3]
```diff
@@ -66,6 +66,9 @@ library DepositUtils {
         uint256 fundedAt; // timestamp when funding proof was received
         bytes utxoOutpoint;  // the 36-byte outpoint of the custodied UTXO
 
+        /// @notice Map of ETH balances an address can withdraw after contract reaches ends-state.
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/DepositUtils.sol:69]: Why a `@notice` here when we don't do it anywhere else?

File: solidity/contracts/deposit/DepositUtils.sol [hunk 4]
```diff
@@ -396,6 +399,34 @@ library DepositUtils {
         return _postCallBalance.sub(_preCallBalance);
     }
 
+    /// @notice     Adds a given amount to an addresses withdrawal allowance.
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/DepositUtils.sol:402]: I would rephrase “Adds a given amount to the withdrawal allowance for the address.” to avoid the awkward address possessive here…

File: solidity/contracts/deposit/DepositUtils.sol [hunk 5]
```diff
@@ -396,6 +399,34 @@ library DepositUtils {
...
+
+        // zero-out to prevent reentrancy
+        _d.withdrawalAllowances[msg.sender] = 0;
+
+        /* solium-disable-next-line security/no-call-value */
+        msg.sender.call.value(available)("");
+    }
+
+    /// @notice     Get the caller's withdraw allowance.
+    /// @return     The caller's withdrawable allowance in wei.
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/DepositUtils.sol:425]: We're back-and-forthing between “withdraw allowance” and “withdrawable allowance”. I'm fine with “withdraw allowance”, but let's use it everywhere.

File: solidity/test/DepositFundingTest.js [hunk 6]
```diff
@@ -218,16 +218,15 @@ describe("DepositFunding", async function() {
...
       const signingGroupRequestedAt = await testDeposit.getSigningGroupRequestedAt.call()
-      const finalFunderBalance = await web3.eth.getBalance(owner)
 
-      expect(
-        new BN(finalFunderBalance).sub(new BN(initialFunderBalance)),
-      ).to.eq.BN(openKeepFee)
+      const withdrawable = await testDeposit.getWithdrawAllowance.call({
+        from: owner,
+      })
+      expect(withdrawable).to.eq.BN(new BN(openKeepFee))
```
Comment (Shadowfiend) [inline solidity/test/DepositFundingTest.js:229]: This is kind of nice. Can we also assert that the deposit balance is equal to this amount?

File: solidity/test/DepositLiquidationTest.js [hunk 7]
```diff
@@ -163,23 +163,25 @@ describe("DepositLiquidation", async function() {
...
+        from: buyer,
+      })
 
-      expect(
-        new BN(finalBalance),
-        "buyer balance should increase",
-      ).to.be.gte.BN(initialBalance)
+      expect(withdrawable, "buyer should have a withdrawable balance").to.eq.BN(
+        auctionValue,
+      )
```
Comment (Shadowfiend) [inline solidity/test/DepositLiquidationTest.js:184]: Ditto here on asserting the deposit has this amount in escrow.

File: solidity/test/DepositLiquidationTest.js [hunk 8]
```diff
@@ -203,24 +202,22 @@ describe("DepositLiquidation", async function() {
...
 
       const totalReward = (value * (100 - basePercentage)) / 100
       const split = totalReward / 2
 
-      expect(new BN(split)).to.eq.BN(initiatorBalanceDiff)
+      const withdrawable = await testDeposit.getWithdrawAllowance.call({
+        from: liquidationInitiator,
+      })
+
+      expect(new BN(split)).to.eq.BN(withdrawable)
```
Comment (Shadowfiend) [inline solidity/test/DepositLiquidationTest.js:220]: \+ again for the deposit balance check.

File: solidity/test/DepositLiquidationTest.js [hunk 9]
```diff
@@ -246,24 +240,22 @@ describe("DepositLiquidation", async function() {
...
-      const initiatorBalanceDiff = new BN(finalInitiatorBalance).sub(
-        new BN(initialInitiatorBalance),
-      )
       const signerBalanceDiff = new BN(finalSignerBalance).sub(
         new BN(initialSignerBalance),
       )
 
+      const withdrawable = await testDeposit.getWithdrawAllowance.call({
+        from: liquidationInitiator,
+      })
```
Comment (Shadowfiend) [inline solidity/test/DepositLiquidationTest.js:253]: Uno mas on the deposit balance check.

File: solidity/test/DepositUtilsTest.js [hunk 10]
```diff
@@ -641,6 +642,71 @@ describe("DepositUtils", async function() {
...
+
+    it("correctly adds value to withdrawalAllowances", async () => {
+      const value = new BN(10000)
+      const beneficiary = accounts[1]
+
+      await testDeposit.enableWithdrawal(accounts[1], value)
+
+      const withdrawable = await testDeposit.getWithdrawAllowance.call({
+        from: beneficiary,
+      })
```
Comment (Shadowfiend) [inline solidity/test/DepositUtilsTest.js:662]: We should probably assert this is 0 before the `enableWithdrawal` call.

File: solidity/test/DepositUtilsTest.js [hunk 11]
```diff
@@ -641,6 +642,71 @@ describe("DepositUtils", async function() {
...
+  describe("withdrawFunds()", async () => {
+    beforeEach(async () => {
+      await createSnapshot()
+    })
+
+    afterEach(async () => {
+      await restoreSnapshot()
+    })
+
+    it("Correctly withdraws withdrawable balance at end-state", async () => {
```
Comment (Shadowfiend) [inline solidity/test/DepositUtilsTest.js:677]: We should make this test check every state that is an end-state, and make the next test check every state that isn't.

Also, as far as I know we don't capitalize our “it” strings. You should consider these the end of a sentence that follows from previous `describe`s (e.g. `DepositUtils withdrawFunds() correctly withdraws withdrawable balance at end-state`).

File: solidity/test/DepositUtilsTest.js [hunk 12]
```diff
@@ -641,6 +642,71 @@ describe("DepositUtils", async function() {
...
+      await testDeposit.enableWithdrawal(beneficiary, value)
+      await testDeposit.withdrawFunds({from: beneficiary})
+
+      const withdrawable = await testDeposit.getWithdrawAllowance.call({
+        from: beneficiary,
+      })
+      const finalBalance = await web3.eth.getBalance(beneficiary)
+
+      expect(withdrawable).to.eq.BN(0)
+      expect(finalBalance).to.gt.BN(initialBalance)
```
Comment (Shadowfiend) [inline solidity/test/DepositUtilsTest.js:693]: Just because the final balance is > initial balance, doesn't mean we succeeded, right? We need to make sure we withdraw initial balance + the value enabled for withdrawal.

File: solidity/contracts/deposit/DepositUtils.sol (hunk 3 above)
Comment (NicholasDotSol) [reply solidity/contracts/deposit/DepositUtils.sol:69]: we do it in `approvedDigests` mapping right below. We can in-line both instead

Comment (Shadowfiend) [reply solidity/contracts/deposit/DepositUtils.sol:69]: I'm down either way---if this results in proper API doc JSON that's fine, though we should use `@dev` rather than `@notice` given I don't think we expect users to directly interact with the mapping.

File: solidity/contracts/deposit/DepositLiquidation.sol (hunk 2 above)
Comment (NicholasDotSol) [reply solidity/contracts/deposit/DepositLiquidation.sol:198]: derp, brain glitched on this, thought about it too... nice catch

File: solidity/test/DepositFundingTest.js (hunk 6 above)
Comment (NicholasDotSol) [reply solidity/test/DepositFundingTest.js:229]: Yeah that would be good

File: solidity/test/DepositLiquidationTest.js (hunk 7 above)
Comment (NicholasDotSol) [reply solidity/test/DepositLiquidationTest.js:184]: In this case, we need `contractBalance = purchesedBond + remainder / 2` since we're not in fraud.

File: solidity/test/DepositUtilsTest.js (hunk 12 above)
Comment (NicholasDotSol) [reply solidity/test/DepositUtilsTest.js:693]: That's right, I got lazy accounting for the gas cost since we're calling from `beneficiary`. Meant to fix this
